File: pwiz_tools/BiblioSpec/src/ShimadzuMLBReader.cpp
//
// $Id: ShimadzuMLBReader.cpp 9898 2016-07-13 22:38:39Z kaipot $
//
//
// Original author: Brian Pratt <bspratt@u.washington.edu>
// Based on msfReader by Kaipo Tamura <kaipot@u.washington.edu>
//
// Copyright 2016 University of Washington - Seattle, WA 98195
//
// Licensed under the Apache License, Version 2.0 (the "License"); 
// you may not use this file except in compliance with the License. 
// You may obtain a copy of the License at 
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software 
// distributed under the License is distributed on an "AS IS" BASIS, 
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
// See the License for the specific language governing permissions and 
// limitations under the License.
//

#include "ShimadzuMLBReader.h"

namespace BiblioSpec
{
    ShimadzuMLBReader::ShimadzuMLBReader(BlibBuilder& maker, 
                         const char* mlbFile, 
                         const ProgressIndicator* parent_progress)
        : BuildParser(maker, mlbFile, parent_progress),
        mlbName_(mlbFile), schemaVersion_(-1)
    {
        setSpecFileName(mlbFile, false);
        lookUpBy_ = INDEX_ID;
        // point to self as spec reader
        delete specReader_;
        specReader_ = this;
    }

    ShimadzuMLBReader::~ShimadzuMLBReader()
    {
        sqlite3_close(mlbFile_);

        specReader_ = NULL; // so the parent class doesn't try to delete itself
        // free spectra
        for (map<int, SpecData*>::iterator it = spectra_.begin(); it != spectra_.end(); ++it)
        {
            if (it->second != NULL)
            {
                delete it->second;
                it->second = NULL;
            }
        }
    }


    bool ShimadzuMLBReader::parseFile()
    {
        sqlite3_open(mlbName_, &mlbFile_);
        if (!mlbFile_)
        {
            throw BlibException(true, "Couldn't open '%s'.", mlbName_);
        }
        // Get the schema version
        sqlite3_stmt* statement = getStmt(
            "SELECT DBVer FROM PROPERTY");
        if (hasNext(statement))
        {
            string version = lexical_cast<string>(sqlite3_column_text(statement, 0));
            sqlite3_finalize(statement);
            vector<string> versionPieces;
            boost::split(versionPieces, version, boost::is_any_of("."));
            try {
                schemaVersion_ = lexical_cast<int>(versionPieces.front());
                Verbosity::debug("Schema version is %d (%s)", schemaVersion_, version.c_str());
            } catch (...) {
                Verbosity::error("Unknown schema version format: '%s'", version.c_str());
            }
        }
        if (schemaVersion_ < 1) {
            Verbosity::error("Could not determine schema version.");
        }
        readMSMSSP();
        // add psms by filename
        for (map< string, vector<PSM*> >::iterator iter = fileMap_.begin();
             iter != fileMap_.end();
             ++iter)
        {
            if (iter->second.size() > 0)
            {
                psms_.assign(iter->second.begin(), iter->second.end());
                setSpecFileName(iter->first, false);
                buildTables(UNKNOWN_SCORE_TYPE, iter->first);
            }
        }

        return true;
    }

    vector<PSM_SCORE_TYPE> ShimadzuMLBReader::getScoreTypes() {
        return vector<PSM_SCORE_TYPE>(1, UNKNOWN_SCORE_TYPE);
    }

    double ShimadzuMLBReader::ReadDoubleFromBuffer(const char * & buf)
    {
        double result;
        char *p = reinterpret_cast<char *>(&result);
        for (int n = sizeof(double); n--;)
            *p++ = *buf++;
        return result;
    }


    const char * ShimadzuMLBReader::getAdduct(int adductType, int& charge)
    {
        const char *precursorAdduct;
        switch (adductType) // Per email from Shimadzu's Yutaro Yamamura to Brian Pratt
        {
        case 0x1:
            precursorAdduct = "[M+H]";
            charge = 1;
            break;
        case 0x2:
            precursorAdduct = "[M+Na]";
            charge = 1;
            break;
        case 0x4:
            precursorAdduct = "[M+K]";
            charge = 1;
            break;
        case 0x8:
            precursorAdduct = "[M+NH4]";
            charge = 1;
            break;
        case 0x10: // "other(+)"
            precursorAdduct = "[M+]";
            charge = 1;
            break;
        case 0x20:
            precursorAdduct = "[M-H]";
            charge = 1;
            break;
        case 0x40:
            precursorAdduct = "[M+HCOO]";
            charge = 1;
            break;
        case 0x80:
            precursorAdduct = "[M+CH3COO]";
            charge = 1;
            break;
        case 0x100:
            precursorAdduct = "[M+Cl]";
            charge = 1;
            break;
        case 0x200: // "other(-)"
            precursorAdduct = "[M-]";
            charge = 1;
            break;
        default:
            precursorAdduct = NULL;
            charge = 0;
            break;
        }
        if (precursorAdduct == NULL)
        {
            throw BlibException(false, "Unknown adduct type");
        }
        return precursorAdduct;
    }

    void ShimadzuMLBReader::readMSMSSP()
    {
        int specCount =
            getRowCount("(SELECT DISTINCT SP_ID FROM MSMSSP WHERE MSStage IS 2)");
        Verbosity::status("Parsing %d spectra.", specCount);
        ProgressIndicator progress(specCount);
        sqlite3_stmt* statement = 
            getStmt(
                "SELECT SP_ID, RT, PrecursorMZ, SpecPeak, AdductType, CompForm, CompName, IUPACNo, CASNo, DataFilePath "
                "FROM MSMSSP "
                "WHERE SP_ID IN (SELECT DISTINCT SP_ID FROM MSMSSP WHERE MSStage IS 2)");

        // turn each row of returned table into a spectrum
        while (hasNext(statement))
        {
            SpecData* specData = new SpecData();
            int specId = sqlite3_column_int(statement, 0);
            specData->id = specId;
            specData->retentionTime = sqlite3_column_double(statement, 1) / 60000.0; // convert msec to minutes
            specData->mz = sqlite3_column_int(statement, 2);
            const char* spectrum_ptr = static_cast<const char*>(sqlite3_column_blob(statement, 3));
            size_t spectrumLen = sqlite3_column_bytes(statement, 3);
            specData->numPeaks = spectrumLen / (4 * sizeof(double));
            specData->mzs = new double[specData->numPeaks];
            specData->intensities = new float[specData->numPeaks];
            int confirm_nPeaks = ReadDoubleFromBuffer(spectrum_ptr)/4;
            if (confirm_nPeaks != specData->numPeaks)
            {
                throw BlibException(false, "Inconsistent peak count");
            }
            for (int n = 0; n <  specData->numPeaks; n++)
            {
                specData->mzs[n] = ReadDoubleFromBuffer(spectrum_ptr);
                specData->intensities[n] = ReadDoubleFromBuffer(spectrum_ptr);
                spectrum_ptr += 2 * sizeof(double);
            }
            int adductType = sqlite3_column_int(statement, 4);
            int charge;

            curPSM_ = new PSM();
            curPSM_->specKey = specId;
            curPSM_->specIndex = specId;
			curPSM_->smallMolMetadata.precursorAdduct = getAdduct(adductType, charge);
            curPSM_->charge = charge;
            curPSM_->smallMolMetadata.chemicalFormula = lexical_cast<string>(sqlite3_column_text(statement, 5));
            curPSM_->smallMolMetadata.moleculeName = lexical_cast<string>(sqlite3_column_text(statement, 6));
            curPSM_->smallMolMetadata.inchiKey = lexical_cast<string>(sqlite3_column_text(statement, 7));
            string cas = lexical_cast<string>(sqlite3_column_text(statement, 8));
            bal::replace_all(cas, " ", ""); // Example had random spaces in the CAS string
            if (!cas.empty())
            {
                curPSM_->smallMolMetadata.otherKeys = "cas:" + cas;
            }

            string datafilepath = lexical_cast<string>(sqlite3_column_text(statement, 9));
            if (datafilepath.empty())
                datafilepath = "unknown";
            fileMap_[datafilepath].push_back(curPSM_);

            // add spectrum to map
            spectra_[specId] = specData;
            spectraChargeStates_[specId] = charge;

            progress.increment();
        }

        Verbosity::debug("Map has %d spectra", spectra_.size());
    }

    /**
     * Prepares the given string as a SQLite query and verifies the return value.
     */
    sqlite3_stmt* ShimadzuMLBReader::getStmt(const string& query) const
    {
        return getStmt(mlbFile_, query);
    }
    
    sqlite3_stmt* ShimadzuMLBReader::getStmt(sqlite3* handle, const string& query) {
        sqlite3_stmt* statement;
        if (sqlite3_prepare(handle, query.c_str(), -1, &statement, NULL) != SQLITE_OK) {
            throw BlibException("Cannot prepare SQL statement: %s ", sqlite3_errmsg(handle));
        }
        return statement;
    }

    /**
     * Attempts to return the next row of the SQLite statement. Returns true on success.
     * Otherwise, finalize the statement and return false.
     */
    bool ShimadzuMLBReader::hasNext(sqlite3_stmt* statement)
    {
        if (sqlite3_step(statement) != SQLITE_ROW) {
            sqlite3_finalize(statement);
            return false;
        }
        return true;
    }

    /**
     * Returns the number of records in the SQLite table.
     */
    int ShimadzuMLBReader::getRowCount(string table) const
    {
        sqlite3_stmt* statement = getStmt("SELECT COUNT(*) FROM " + table);

        sqlite3_step(statement);
        int count = sqlite3_column_int(statement, 0);
        sqlite3_finalize(statement);

        return count;
    }

    // SpecFileReader methods
    /**
     * Implemented to satisfy SpecFileReader interface.  Since spec and
     * results files are the same, no need to open a new one.
     */
    void ShimadzuMLBReader::openFile(const char* filename, bool mzSort) {}

    void ShimadzuMLBReader::setIdType(SPEC_ID_TYPE type) {}

    /**
     * Return a spectrum via the returnData argument.  If not found in the
     * spectra map, return false and leave returnData unchanged.
     */
    bool ShimadzuMLBReader::getSpectrum(int identifier, SpecData& returnData, SPEC_ID_TYPE findBy, bool getPeaks)
    {
        map<int, SpecData*>::iterator found = spectra_.find(identifier);
        if (found == spectra_.end())
        {
            return false;
        }

        SpecData* foundData = found->second;
        returnData = *foundData;
        return true;
    }

    /**
     * Only specific spectra can be accessed from the ShimadzuMLBReader.
     */
    bool ShimadzuMLBReader::getSpectrum(string identifier, SpecData& returnData, bool getPeaks)
    {
        Verbosity::warn("ShimadzuMLBReader cannot fetch spectra by string identifier, "
                        "only by spectrum index.");
        return false;
    }

    /**
     * Only specific spectra can be accessed from the ShimadzuMLBReader.
     */
    bool ShimadzuMLBReader::getNextSpectrum(SpecData& returnData, bool getPeaks)
    {
        Verbosity::warn("ShimadzuMLBReader does not support sequential file reading.");
        return false;
    }
}

File: pwiz-sharp/Tools/BiblioSpec/src/BiblioSpec/ShimadzuMLBReader.cs
// Port of pwiz_tools/BiblioSpec/src/ShimadzuMLBReader.{h,cpp}
//
// Parses Shimadzu LabSolutions small-molecule libraries (.mlb), which are SQLite databases
// with a single MSMSSP table holding both PSM metadata and the spectrum peak blobs. The
// reader doubles as the spec-file reader (cpp parity: `specReader_ = this`) — peaks are
// looked up out of the same SQLite handle by the spectrum's index id.

using System.Data.SQLite;
using System.Globalization;

namespace Pwiz.Tools.BiblioSpec;

/// <summary>
/// Parses Shimadzu LabSolutions <c>.mlb</c> small-molecule libraries. cpp parity:
/// <c>BiblioSpec::ShimadzuMLBReader</c> at
/// <c>pwiz_tools/BiblioSpec/src/ShimadzuMLBReader.{h,cpp}</c>.
/// </summary>
/// <remarks>
/// <para>The .mlb file is a SQLite database, NOT a vendor-SDK format. The reader opens the
/// db once, iterates MSMSSP rows where MSStage==2, and builds one <see cref="PSM"/> per
/// distinct SP_ID. The spectrum peaks ship as a binary blob in the same row, so the reader
/// also serves as the <see cref="ISpecFileReader"/>: cpp sets <c>specReader_ = this</c> in
/// its ctor; the C# port does the same by installing an inner reader that reads back from
/// an in-memory <see cref="Dictionary{TKey,TValue}"/> populated during
/// <see cref="ParseFile"/>.</para>
/// <para>Each row's DataFilePath column groups PSMs by source data-file; cpp emits one
/// SpectrumSourceFile row per group via <see cref="BuildParser.SetSpecFileName(string,bool)"/>
/// followed by <see cref="BuildParser.BuildTables(PsmScoreType,string,bool,WorkflowType)"/>.</para>
/// </remarks>
public sealed class ShimadzuMLBReader : BuildParser
{
    private readonly string _mlbName;
    private SQLiteConnection? _mlbFile;
    private int _schemaVersion = -1;

    // cpp parity: ShimadzuMLBReader.h:47 — DataFilePath -> PSMs that came out of that file.
    private readonly Dictionary<string, List<PSM>> _fileMap = new(StringComparer.Ordinal);

    // cpp parity: ShimadzuMLBReader.h:49 — SP_ID -> in-memory SpecData (the peak list comes
    // out of the same SQLite row that produced the PSM, so we keep it cached here for the
    // later spectrum lookups from BuildParser.BuildTables).
    private readonly Dictionary<int, SpecData> _spectra = new();

    /// <summary>Returns true if <paramref name="path"/> ends with <c>.mlb</c>.</summary>
    public static bool AcceptsExtension(string path) =>
        path.EndsWith(".mlb", StringComparison.OrdinalIgnoreCase);

    /// <summary>
    /// Construct a ShimadzuMLBReader bound to <paramref name="builder"/> and the file at
    /// <paramref name="mlbFile"/>. cpp parity: ShimadzuMLBReader.cpp:27.
    /// </summary>
    public ShimadzuMLBReader(BlibBuilder builder, string mlbFile, ProgressIndicator? parentProgress)
        : base(builder, mlbFile, parentProgress)
    {
        _mlbName = mlbFile;

        // cpp parity: ShimadzuMLBReader.cpp:33 — record the mlb path as the spec-file, no
        // existence check (we already know we opened it).
        SetSpecFileName(mlbFile, checkFile: false);

        // cpp parity: ShimadzuMLBReader.cpp:34 — spectra are addressed by their SP_ID, which
        // we treat as an index id.
        LookUpBy = SpecIdType.IndexId;

        // cpp parity: ShimadzuMLBReader.cpp:37 — point the spec reader at ourselves via an
        // inner adapter that reads back from _spectra.
        SpecReader = new MlbSpecFileReader(this);
    }

    /// <summary>Score types this reader produces. cpp parity: ShimadzuMLBReader.cpp:100.</summary>
    public override IList<PsmScoreType> GetScoreTypes() =>
        new[] { PsmScoreType.UnknownScoreType };

    /// <summary>
    /// Open the SQLite db, read its schema version, parse all MSMSSP MS2 rows into
    /// <see cref="_fileMap"/> + <see cref="_spectra"/>, then emit one library entry per
    /// DataFilePath group. cpp parity: ShimadzuMLBReader.cpp:57.
    /// </summary>
    public override bool ParseFile()
    {
        // cpp parity: ShimadzuMLBReader.cpp:59 — open the database; on failure cpp throws.
        // System.Data.SQLite throws on Open(); we wrap into a BlibException for parity.
        var connStr = new SQLiteConnectionStringBuilder
        {
            DataSource = _mlbName,
            ReadOnly = true,
        }.ToString();

        try
        {
            _mlbFile = new SQLiteConnection(connStr);
            _mlbFile.Open();
        }
        catch (Exception ex)
        {
            throw new BlibException(true, $"Couldn't open '{_mlbName}': {ex.Message}");
        }

        // cpp parity: ShimadzuMLBReader.cpp:64 — read the schema version from PROPERTY.
        using (var cmd = _mlbFile.CreateCommand())
        {
            cmd.CommandText = "SELECT DBVer FROM PROPERTY";
            using var reader = cmd.ExecuteReader();
            if (reader.Read())
            {
                var version = reader.IsDBNull(0) ? string.Empty : reader.GetValue(0)?.ToString() ?? string.Empty;
                var pieces = version.Split('.');
                if (pieces.Length > 0 &&
                    int.TryParse(pieces[0], NumberStyles.Integer, CultureInfo.InvariantCulture, out var major))
                {
                    _schemaVersion = major;
                    Verbosity.Debug($"Schema version is {_schemaVersion} ({version})");
                }
                else
                {
                    Verbosity.Error($"Unknown schema version format: '{version}'");
                }
            }
        }

        if (_schemaVersion < 1)
        {
            Verbosity.Error("Could not determine schema version.");
        }

        ReadMsmsSp();

        // cpp parity: ShimadzuMLBReader.cpp:84 — flush each DataFilePath group to the library.
        foreach (var kvp in _fileMap)
        {
            if (kvp.Value.Count == 0) continue;

            Psms.Clear();
            foreach (var psm in kvp.Value)
                Psms.Add(psm);

            SetSpecFileName(kvp.Key, checkFile: false);
            BuildTables(PsmScoreType.UnknownScoreType, kvp.Key);
        }

        return true;
    }

    /// <summary>
    /// cpp parity: ShimadzuMLBReader.cpp:171 — iterate every MS2 row in MSMSSP, building
    /// one PSM + cached SpecData per SP_ID.
    /// </summary>
    private void ReadMsmsSp()
    {
        if (_mlbFile is null)
            throw new InvalidOperationException("MLB file not opened.");

        int specCount = GetRowCount(
            "(SELECT DISTINCT SP_ID FROM MSMSSP WHERE MSStage IS 2)");
        Verbosity.Status($"Parsing {specCount.ToString(CultureInfo.InvariantCulture)} spectra.");
        var progress = new ProgressIndicator(specCount);

        using var cmd = _mlbFile.CreateCommand();
        cmd.CommandText =
            "SELECT SP_ID, RT, PrecursorMZ, SpecPeak, AdductType, CompForm, CompName, IUPACNo, CASNo, DataFilePath " +
            "FROM MSMSSP " +
            "WHERE SP_ID IN (SELECT DISTINCT SP_ID FROM MSMSSP WHERE MSStage IS 2)";

        using var reader = cmd.ExecuteReader();
        while (reader.Read())
        {
            var specData = new SpecData();
            int specId = ReadIntColumn(reader, 0);
            specData.Id = specId;

            // cpp parity: ShimadzuMLBReader.cpp:189 — RT is msec; library wants minutes.
            specData.RetentionTime = ReadDoubleColumn(reader, 1) / 60000.0;

            // cpp parity: ShimadzuMLBReader.cpp:190 — cpp uses sqlite3_column_int on the
            // PrecursorMZ column; preserve the truncating-to-int semantics.
            specData.Mz = ReadIntColumn(reader, 2);

            // cpp parity: ShimadzuMLBReader.cpp:191-206 — SpecPeak is a binary blob laid out
            // as a sequence of doubles: [peakCount*4, then for each peak (mz, intensity,
            // 2-double pad)]. peakCount = blob_bytes / (4 * sizeof(double)).
            var blob = (byte[])reader.GetValue(3);
            int numPeaks = blob.Length / (4 * sizeof(double));
            specData.NumPeaks = numPeaks;
            specData.Mzs = new double[numPeaks];
            specData.Intensities = new float[numPeaks];

            int offset = 0;
            // First double in the blob is the entry count (peaks*4); used as a sanity check.
            double headerCount = ReadDoubleFromBuffer(blob, ref offset);
            int confirmNPeaks = (int)(headerCount / 4);
            if (confirmNPeaks != numPeaks)
            {
                throw new BlibException(false, "Inconsistent peak count");
            }
            for (int n = 0; n < numPeaks; n++)
            {
                specData.Mzs[n] = ReadDoubleFromBuffer(blob, ref offset);
                // cpp parity: ShimadzuMLBReader.cpp:204 — intensity is the second double of
                // the per-peak quad, then 2 doubles of padding (skipped).
                specData.Intensities[n] = (float)ReadDoubleFromBuffer(blob, ref offset);
                offset += 2 * sizeof(double);
            }

            int adductType = ReadIntColumn(reader, 4);
            string precursorAdduct = GetAdduct(adductType, out int charge);

            var psm = new PSM
            {
                SpecKey = specId,
                SpecIndex = specId,
                Charge = charge,
            };
            psm.SmallMolMetadata.PrecursorAdduct = precursorAdduct;
            psm.SmallMolMetadata.ChemicalFormula = ReadStringColumn(reader, 5);
            psm.SmallMolMetadata.MoleculeName = ReadStringColumn(reader, 6);
            psm.SmallMolMetadata.InchiKey = ReadStringColumn(reader, 7);

            // cpp parity: ShimadzuMLBReader.cpp:218 — strip stray spaces in the CAS column.
            var cas = ReadStringColumn(reader, 8).Replace(" ", string.Empty);
            if (cas.Length > 0)
            {
                psm.SmallMolMetadata.OtherKeys = "cas:" + cas;
            }

            string dataFilePath = ReadStringColumn(reader, 9);
            if (string.IsNullOrEmpty(dataFilePath))
                dataFilePath = "unknown";

            if (!_fileMap.TryGetValue(dataFilePath, out var list))
            {
                list = new List<PSM>();
                _fileMap[dataFilePath] = list;
            }
            list.Add(psm);

            _spectra[specId] = specData;

            progress.Increment();
        }

        Verbosity.Debug($"Map has {_spectra.Count.ToString(CultureInfo.InvariantCulture)} spectra");
    }

    /// <summary>cpp parity: ShimadzuMLBReader.cpp:104 — read 8 bytes from <paramref name="buf"/>
    /// at <paramref name="offset"/> as a little-endian double. Advances <paramref name="offset"/>.</summary>
    private static double ReadDoubleFromBuffer(byte[] buf, ref int offset)
    {
        double v = BitConverter.ToDouble(buf, offset);
        offset += sizeof(double);
        return v;
    }

    /// <summary>cpp parity: ShimadzuMLBReader.cpp:114 — Shimadzu adduct-type bitmask
    /// translation, courtesy of Yutaro Yamamura via Brian Pratt.</summary>
    private static string GetAdduct(int adductType, out int charge)
    {
        switch (adductType)
        {
            case 0x1:   charge = 1; return "[M+H]";
            case 0x2:   charge = 1; return "[M+Na]";
            case 0x4:   charge = 1; return "[M+K]";
            case 0x8:   charge = 1; return "[M+NH4]";
            case 0x10:  charge = 1; return "[M+]";       // other(+)
            case 0x20:  charge = 1; return "[M-H]";
            case 0x40:  charge = 1; return "[M+HCOO]";
            case 0x80:  charge = 1; return "[M+CH3COO]";
            case 0x100: charge = 1; return "[M+Cl]";
            case 0x200: charge = 1; return "[M-]";       // other(-)
            default:
                charge = 0;
                throw new BlibException(false, "Unknown adduct type");
        }
    }

    /// <summary>Read a string column safely; returns empty string when DBNull.</summary>
    private static string ReadStringColumn(SQLiteDataReader reader, int ordinal)
    {
        if (reader.IsDBNull(ordinal)) return string.Empty;
        var value = reader.GetValue(ordinal);
        return value?.ToString() ?? string.Empty;
    }

    /// <summary>Read an int column tolerantly — cpp's sqlite3_column_int coerces text and reals;
    /// System.Data.SQLite's <see cref="SQLiteDataReader.GetInt32"/> is type-strict and would throw
    /// on REAL columns like Shimadzu's PrecursorMZ.</summary>
    private static int ReadIntColumn(SQLiteDataReader reader, int ordinal)
    {
        if (reader.IsDBNull(ordinal)) return 0;
        var value = reader.GetValue(ordinal);
        return value switch
        {
            long l => (int)l,
            int i => i,
            double d => (int)d,
            float f => (int)f,
            decimal m => (int)m,
            string s => int.TryParse(s, NumberStyles.Float, CultureInfo.InvariantCulture, out var parsed)
                ? parsed
                : (int)double.Parse(s, NumberStyles.Float, CultureInfo.InvariantCulture),
            _ => Convert.ToInt32(value, CultureInfo.InvariantCulture),
        };
    }

    /// <summary>Read a double column tolerantly — Shimadzu sometimes stores numeric columns
    /// as TEXT; <see cref="SQLiteDataReader.GetDouble"/> would throw on those.</summary>
    private static double ReadDoubleColumn(SQLiteDataReader reader, int ordinal)
    {
        if (reader.IsDBNull(ordinal)) return 0;
        var value = reader.GetValue(ordinal);
        return value switch
        {
            double d => d,
            float f => f,
            long l => l,
            int i => i,
            decimal m => (double)m,
            string s => double.Parse(s, NumberStyles.Float, CultureInfo.InvariantCulture),
            _ => Convert.ToDouble(value, CultureInfo.InvariantCulture),
        };
    }

    /// <summary>cpp parity: ShimadzuMLBReader.cpp:272 — SELECT COUNT(*) helper.</summary>
    private int GetRowCount(string table)
    {
        if (_mlbFile is null) return 0;
        using var cmd = _mlbFile.CreateCommand();
        cmd.CommandText = "SELECT COUNT(*) FROM " + table;
        var result = cmd.ExecuteScalar();
        return result is null or DBNull ? 0 : Convert.ToInt32(result, CultureInfo.InvariantCulture);
    }

    /// <inheritdoc/>
    protected override void Dispose(bool disposing)
    {
        if (disposing)
        {
            _mlbFile?.Dispose();
            _mlbFile = null;
        }
        base.Dispose(disposing);
    }

    // --- Inner spec reader ---------------------------------------------------------------

    /// <summary>
    /// Inner ISpecFileReader that hands out cached SpecData from the parent reader. cpp
    /// parity: ShimadzuMLBReader.cpp:288-326 — getSpectrum(int, ...) hits the SP_ID -> SpecData
    /// map; the string-id and next-spectrum overloads warn-and-return-false.
    /// </summary>
    private sealed class MlbSpecFileReader : SpecFileReaderBase
    {
        private readonly ShimadzuMLBReader _owner;

        public MlbSpecFileReader(ShimadzuMLBReader owner)
        {
            _owner = owner;
        }

        public override void OpenFile(string path, bool mzSort = false)
        {
            // cpp parity: ShimadzuMLBReader.cpp:288 — no-op (spec and results files are the same).
        }

        public override SpecIdType IdType
        {
            set { /* no-op, cpp parity: ShimadzuMLBReader.cpp:290 */ }
        }

        public override bool GetSpectrum(int identifier, SpecData returnData, SpecIdType findBy, bool getPeaks = true)
        {
            ArgumentNullException.ThrowIfNull(returnData);
            if (!_owner._spectra.TryGetValue(identifier, out var found))
                return false;
            // cpp parity: ShimadzuMLBReader.cpp:305 `returnData = *foundData` — deep-copy via
            // SpecData.CopyFrom (re-allocates the peak arrays).
            returnData.CopyFrom(found);
            return true;
        }

        public override bool GetSpectrum(string identifier, SpecData returnData, bool getPeaks = true)
        {
            Verbosity.Warn(
                "ShimadzuMLBReader cannot fetch spectra by string identifier, only by spectrum index.");
            return false;
        }

        public override bool GetNextSpectrum(SpecData returnData, bool getPeaks = true)
        {
            Verbosity.Warn("ShimadzuMLBReader does not support sequential file reading.");
            return false;
        }
    }
}
